homeassistant/components/wled/__init__.py

```diff
@@ -0,0 +1,181 @@
+    ATTR_IDENTIFIERS,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    ATTR_SOFTWARE_VERSION,
+    DATA_WLED_CLIENT,
+    DATA_WLED_TIMER,
+    DATA_WLED_UPDATED,
+    DOMAIN,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_NAME, CONF_HOST
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+SCAN_INTERVAL = timedelta(seconds=5)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the WLED components."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up WLED from a config entry."""
+
+    # Create WLED instance for this entry
+    session = async_get_clientsession(hass)
+    wled = WLED(entry.data[CONF_HOST], loop=hass.loop, session=session)
+
+    # Ensure we can connect and talk to it
+    try:
+        await wled.update()
+    except WLEDConnectionError as exception:
+        raise ConfigEntryNotReady from exception
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {DATA_WLED_CLIENT: wled}
+
+    # Set up all platforms for this device/entry.
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, LIGHT_DOMAIN)
+    )
+
+    async def interval_update(now: dt_util.dt.datetime = None) -> None:
+        """Poll WLED device function, dispatches event after update."""
+        try:
+            await wled.update()
+        except WLEDError:
+            _LOGGER.debug("An error occurred while updating WLED", exc_info=True)
+
+        # Even if the update failed, we still send out the event.
+        # To allow entities to make themselves unavailable.
+        async_dispatcher_send(hass, DATA_WLED_UPDATED, entry.entry_id)
+
+    # Schedule update interval
+    hass.data[DOMAIN][entry.entry_id][DATA_WLED_TIMER] = async_track_time_interval(
+        hass, interval_update, SCAN_INTERVAL
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload WLED config entry."""
+
+    # Cancel update timer for this entry/device.
+    cancel_timer = hass.data[DOMAIN][entry.entry_id][DATA_WLED_TIMER]
+    cancel_timer()
+
+    # Unload entities for this entry/device.
+    await hass.config_entries.async_forward_entry_unload(entry, LIGHT_DOMAIN)
+
+    # Cleanup
+    del hass.data[DOMAIN][entry.entry_id]
+    if not hass.data[DOMAIN]:
+        del hass.data[DOMAIN]
```

Comment:
This reads weird. besides, afaik, del has no side-effects on none, so doing `del hass.data[DOMAIN]` should be just fine, right?

Reply:
This integration supports multiple configs entries, this check makes sure the domain namespace gets cleaned up after the last entry has been unloaded.